core/spoofax.compiler/src/main/resources/mb/spoofax/compiler/adapter/constraint_analyzer/AnalyzeMultiTaskDef.java.mustache

```diff
@@ -71,18 +71,40 @@ public class {{baseAnalyzeMultiTaskDef.id}} extends ConstraintAnalyzeMultiTaskDe
         context.require(classLoaderResources.tryGetAsNativeResource({{languageProjectInput.constraintAnalyzer.qualifiedId}}.class), ResourceStampers.hashFile());
         context.require(classLoaderResources.tryGetAsNativeResource({{languageProjectInput.constraintAnalyzerFactory.qualifiedId}}.class), ResourceStampers.hashFile());
         final StrategoRuntime strategoRuntime = context.require(getStrategoRuntimeProvider, None.instance).getValue().get();
-        return constraintAnalyzer.analyze(root, asts, constraintAnalyzerContext, strategoRuntime{{#languageProjectInput.enableStatix}}.addContextObject(mb.statix.spoofax.IStatixProjectConfig.class, createStatixProjectConfig()){{/languageProjectInput.enableStatix}}, resourceService);
+        return constraintAnalyzer.analyze(root, asts, constraintAnalyzerContext, decorateStrategoRuntime(strategoRuntime, root), resourceService);
     }
 
+    {{^languageProjectInput.enableStatix}}
+    protected StrategoRuntime decorateStrategoRuntime(StrategoRuntime strategoRuntime, ResourcePath root) {
+        return strategoRuntime;
+    }
+    {{/languageProjectInput.enableStatix}}
     {{#languageProjectInput.enableStatix}}
+    protected StrategoRuntime decorateStrategoRuntime(StrategoRuntime strategoRuntime, ResourcePath root) {
+        return strategoRuntime.addContextObject(mb.statix.spoofax.IStatixProjectConfig. class, createStatixProjectConfig())
+                    .addContextObject(mb.spoofax2.common.primitive.generic.Spoofax2ProjectContext.class, createProjectContext(root));
+    }
+
     protected mb.statix.spoofax.IStatixProjectConfig createStatixProjectConfig() {
         return new mb.statix.spoofax.StatixProjectConfig(
-            null,
+            // Spoofax 2 uses language name here. In Spoofax 3, we use the FQ artifact id, because that can be retrieved from the context object.
+            // Currently, we do not differentiate by version, because that is optional (which slightly complicates creating identical strings.
+            // This place should be kept in sync with `StatixCommonPrimitive`.
+            java.util.Collections.singletonMap("{{qualifiedLanguageId}}", mb.statix.spoofax.SolverMode.{{statixSolverMode}}),
```

Comment:
This is a rather obtuse way to pass the solver mode to the primitives that require it. A nicer way would be to create a new `Spoofax3StatixProjectConfig` class with the solver mode, add that as a context object, and then just retrieve that in the primitive. That avoids the whole messing around with the language ID in `StatixCommonPrimitive`.

However, I am fine with merging it like this to keep it in sync with Spoofax 2.